src/cloudformation_cli_python_lib/recast.py

```diff
@@ -0,0 +1,131 @@
+import typing
+from typing import Any, Dict, List, Mapping, Set
+
+from .exceptions import InvalidRequest
+
+
+# CloudFormation recasts all primitive types as strings, this tries to set them back to
+# the types in the type hints
+def recast_object(
+    cls: Any, json_data: Mapping[str, Any], classes: Dict[str, Any]
+) -> None:
+    if not isinstance(json_data, dict):
+        raise InvalidRequest(f"Can only parse dict items, not {type(json_data)}")
+    # if type is Any, we leave it as is
+    if cls == typing.Any:
+        return
+    for k, v in json_data.items():
+        if isinstance(v, dict):
+            child_cls = _field_to_type(cls.__dataclass_fields__[k].type, k, classes)
```

Comment:
when is this not Any? just trying to get a grasp on logic here

Reply:
wherever we have a strongly defined schema, like `ResourceModel` or [`ADict`](https://github.com/jaymccon/cloudformation-cli-python-plugin/blob/nested_schema_and_recast/tests/lib/sample_model.py#L194-L197)
`Any` only occurs in the case of a dict defined in the schema by `patternProperties`, like [`AFreeformDict`](https://github.com/jaymccon/cloudformation-cli-python-plugin/blob/nested_schema_and_recast/tests/lib/sample_model.py#L43)